LifetimeDependence: Support closures and function types

aidan-hall opened the pull request:
rdar://130303536
rdar://160894371
rdar://166912068
<!--
If this pull request is targeting a release branch, please fill out the
following form:
https://github.com/swiftlang/.github/blob/main/PULL_REQUEST_TEMPLATE/release.md?plain=1

Otherwise, replace this comment with a description of your changes and
rationale. Provide links to external references/discussions if appropriate.
If this pull request resolves any GitHub issues, link them like so:

  Resolves <link to issue>, resolves <link to another issue>.

For more information about linking a pull request to an issue, see:
https://docs.github.com/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
-->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

aidan-hall committed:
LifetimeDependence: Support closures and function types

Closure and type lifetime info can be queried using the request evaluator.

Sema: Diagnose use of disabled ClosureLifetimes feature

aidan-hall committed:
LifetimeDependence: Tests for function type attribute

Includes:
- rdar://166912068 test case
- Nested type SIL tests
- rdar://160894371 test

meg-gupta reviewed (commented):
Please add tests for: 

- SIL (parsing, diagnostics)
- AST printing
- Serialization/Deserialization 

You will also need a new experimental feature flag to guard printing in `.swiftinterface` files since old compilers cannot parse @_lifetime on the function type.

aidan-hall commented on lib/AST/LifetimeDependence.cpp:
```diff
@@ -631,48 +738,48 @@ class LifetimeDependenceChecker {
   bool useLazyInference() const {
-    return isInterfaceFile()
-      || ctx.LangOpts.EnableExperimentalLifetimeDependenceInference;
+    return isInterfaceFile() ||
```
Will revert these changes.
I have been using the `.clang-format` file in the Swift repository; if that doesn't reflect the preferred style, we should update it.

meg-gupta replied:
It's not necessary to revert it in either case.

aidan-hall committed:
[AST] Support multiple of the same attribute per type

aidan-hall committed:
[AST] Print function type LifetimeDependenceInfo using parameter labels

aidan-hall commented:
This PR has rebased on top of #86386

aidan-hall committed:
Serialization: Add LifetimeDependenceInfo.isInferred to module format

In Swift code, explicit lifetime annotations should only ever refer to
parameters by their (internal or external) label. Inferred lifetime
dependencies (for function types) may have targets or sources that have no
label. This means we can only reliably print explicit lifetime annotations,
since all their source and target parameters must have labels.

This works out nicely, since we should always get the same inferred lifetimes,
given the same function interface, so we do not need to include them in module
interface files.

aidan-hall committed:
LifetimeDependence: Fix issues from earlier refactor

This addresses feedback from #86560:
- Remove redundant collectParameterInfo call
- Consistent naming of ParamInfo variables
- Explicit type name instead of decltype

Also add nullptr checks on afd, as appropriate.

aidan-hall committed:
Types: + AnyFunctionType::hasExplicitLifetimeDependencies

aidan-hall committed:
Add experimental feature ClosureLifetimes

aidan-hall committed:
LifetimeDependence: Basic Sema tests for closures

aidan-hall committed:
LifetimeDependence: Use computed closure parameters

Like the result type, the types of the parameters in the closure's parameter
list are not sufficient to perform lifetime inference in many cases.
